pkg/roll/run_update.go

```diff
@@ -360,84 +372,119 @@ func (u *update) countHealthy(id rcf.ID, checks map[string]health.Result) (rcNod
 	}
 
 	afterDelayNew, err := u.countHealthy(u.NewRC, checks)
 	if err != nil {
-		return 0, util.Errorf("Could not determine new service health: %v", err)
+		return 0, 0, util.Errorf("Could not determine new service health: %v", err)
 	}
 
 	afterDelayOld, err := u.countHealthy(u.OldRC, checks)
 	if err != nil {
-		return 0, util.Errorf("Could not determine old service health: %v", err)
+		return 0, 0, util.Errorf("Could not determine old service health: %v", err)
 	}
 
-	afterDelayNext := rollAlgorithm(afterDelayOld.Healthy, afterDelayNew.Healthy, u.DesiredReplicas, u.MinimumReplicas)
+	// We conservatively treat Unknown nodes as healthy on the old side.
+	afterDelayRemove, afterDelayNext := rollAlgorithm(
+		afterDelayOld.Healthy+afterDelayOld.Unknown, afterDelayNew.Healthy,
+		afterDelayOld.Desired+afterDelayNew.Desired, u.DesiredReplicas,
+		u.MinimumReplicas,
+	)
 
-	if afterDelayNext <= 0 {
-		return 0, util.Errorf("No nodes can be safely updated after %v roll delay, will wait again", u.RollDelay)
+	if afterDelayRemove <= 0 && afterDelayNext <= 0 {
+		return 0, 0, util.Errorf("No nodes can be safely updated after %v roll delay, will wait again", u.RollDelay)
 	}
 
-	return afterDelayNext, nil
+	return afterDelayRemove, afterDelayNext, nil
 }
 
-// the roll algorithm defines how to mutate RCs over time. it takes four args:
+// the roll algorithm defines how to mutate RCs over time. it takes five args:
 // - old: the number of healthy nodes on the old RC
 // - new: the number of healthy nodes on the new RC
-// - desired: the number of nodes desired on the new RC (ie the target)
+// - currentDesired: the number of nodes currently desired in total between the old and new RCs
+// - targetDesired: the number of nodes desired on the new RC (ie the target) in the final state
 // - minHealthy: the number of nodes that must always be up (ie the minimum)
-// given these four arguments, rollAlgorithm returns the number of nodes to add
+// given these five arguments, rollAlgorithm returns the number of nodes to add
 // to the new RC and to delete from the old
 //
-// returns zero under the following circumstances:
+// Returns two values, both of which are a positive number or zero:
+// The first value indicates how many nodes to remove from the old RC.
+// The second value indicates how many nodes to add to the old RC.
+//
+// Under the following circumstances, both return values will be zero:
 // - new >= desired (the update is done)
 // - old+new <= minHealthy (at or below the minimum, update has to block)
-func rollAlgorithm(old, new, desired, minHealthy int) int {
+func rollAlgorithm(old, new, currentDesired, targetDesired, minHealthy int) (int, int) {
 	// how much "headroom" do we have between the number of nodes that are
 	// currently healthy, and the number that must be healthy?
 	// if we schedule more than this, we'll go below the minimum
 	headroom := old + new - minHealthy
```

Comment:
add comment that this can go negative! and then `headroom += capacityIncrease` may increase that to just above positive, not the full value of `capacityIncrease`